shared/profile/pgp/index.js

```diff
@@ -0,0 +1,157 @@
+// @flow
+import ProvePgpChoice from './prove-pgp-choice'
+import ImportPgp from './prove-pgp-import'
+import GeneratePgp from './generating-pgp'
+import Finished from './finished-generating-pgp'
+import PgpInfo from './add'
+import React, {Component} from 'react'
+import {Map} from 'immutable'
+import {TypedConnector} from '../../util/typed-connect'
+import {updatePgpInfo, generatePgp} from '../../actions/profile'
+import {navigateUp, navigateTo, routeAppend} from '../../actions/router'
+import * as Constants from '../../constants/profile'
+
+import type {Props as ProvePgpChoiceProps} from './prove-pgp-choice'
+import type {Props as InfoProps} from './add'
+import type {Props as GenerateProps} from './generating-pgp'
+import type {Props as FinishedProps} from './finished-generating-pgp'
+import type {Props as ImportProps} from './prove-pgp-import'
+import type {TypedDispatch} from '../../constants/types/flux'
+import type {TypedState} from '../../constants/reducer'
+import type {URI} from '../../constants/router'
+
+type PgpFlowProps = {
+  phase: 'choice',
+  props: ProvePgpChoiceProps,
+} | {
+  phase: 'import',
+  props: ImportProps,
+} | {
+  phase: 'provideInfo',
+  props: InfoProps,
+} | {
+  phase: 'generate',
+  props: GenerateProps,
+} | {
+  phase: 'finished',
+  props: FinishedProps,
+}
+
+type PgpPhases = 'choice' | 'import' | 'generate' | 'provideInfo' | 'finished'
+// TODO (MM): It's possible to be more explicit here.
+// For example we could say we only support import route
+// iff it came from choice (choice -> import)
+// It might look like [['choice', ['import', ['provideInfo', 'generate']]]
+const handledRoutes: Array<PgpPhases> = ['choice', 'import', 'provideInfo', 'generate', 'finished']
+
+function pathToPhase (currentPath: Map<string, string>, uri: URI): ?PgpPhases {
+  const path = currentPath.get('path') || ''
+  // A bit akward, but this is the type safe way. (read: flow is cool with it)
+  const phase = handledRoutes.find(s => s === path)
+  if (!phase) {
+    console.warn(`Pgp Flow Container can not handle ${path}`)
+    return null
+  }
+  return phase
+}
+
+function pgpRouter (currentPath: Map<string, string>, uri: URI): any {
+  const phase = pathToPhase(currentPath, uri)
+  if (!phase) {
+    throw new Error('Pgp Router failed', currentPath, uri)
+  }
+
+  return {
+    componentAtTop: {
+      element: <ConnectedPgpFlowContainer phase={phase} />,
+      hideBack: true,
+      hideNavBar: true,
+    },
+    parseNextRoute: pgpRouter,
+  }
+}
+
+// Handle the flow between the dumb components
+class PgpFlowContainer extends Component<void, PgpFlowProps, void> {
+  render () {
+    switch (this.props.phase) {
```

Comment:
still not clear to me the phase concept is better than just regular routing. like why is this component a mini router instead of us just navigating?